WB-1051 - Add States to TextField

RivasCVA opened the pull request:
## Summary:
Integrated the States that TextField will use. This includes `enabled`,
`disabled`, `error`, and `focused`. Included the `focused` style for
when the user is focused on (editing) the TextField. I attached images
below of this new style.

Issue: https://khanacademy.atlassian.net/browse/WB-1051

## Test plan:
1. View the generated Storybook link. Check the "Form / TextField" section.
2. You can also view Styleguidist under the "Form" section.
###### NOTE: No props have been added yet so the TextField is stuck in the `enabled` state, but it will also show the `focused` state when you click on the TextField.

##### onFocus [no text]
![focus-notext](https://user-images.githubusercontent.com/60367213/119385979-89e71c00-bc8c-11eb-92bd-04f9931cbbd1.png)

##### onFocus [text]
![focus-text](https://user-images.githubusercontent.com/60367213/119386006-93708400-bc8c-11eb-86e1-9d1a5a99081c.png)

jandrade commented on packages/wonder-blocks-form/src/components/text-field.js:
```diff
@@ -7,16 +7,48 @@ import Color from "@khanacademy/wonder-blocks-color";
 
-export default class TextField extends React.Component<{||}> {
+type State = {|
+    enabled: boolean,
```
In this case it's better to have only one variable that holds the `enabled/disabled` state. There could be situations for X reasons where both could be `true` and that would lead to an invalid state. In that case it's better to get rid of `enabled` and use only `disabled=true|false` instead.

RivasCVA replied:
Yes. That makes total sense. I was a bit confused on why the spec had both when I started to code this up. I went too deep into this and just thought that I would have to internally ensure that `enabled` and `disabled` were never true at the same time. This clears it up 😄

jandrade replied:
yeah no worries, sometimes there will be some stuff that doesn't translate exactly from design to UI :). And I think that's one of the reasons why we use implementation specs, so we can create a bridge between Design and Code.

jandrade commented on packages/wonder-blocks-form/src/components/text-field.js:
```diff
@@ -7,16 +7,48 @@ import Color from "@khanacademy/wonder-blocks-color";
-export default class TextField extends React.Component<{||}> {
+type State = {|
+    enabled: boolean,
+    disabled: boolean,
```
Based on the previous comment, `disabled` should be treated as a prop that is only defined in the component instance and shouldn't change internally (only via the `disabled` prop).

jandrade commented on packages/wonder-blocks-form/src/components/text-field.js:
```diff
@@ -7,16 +7,48 @@ import Color from "@khanacademy/wonder-blocks-color";
+        focused: false,
+    };
+
+    handleOnFocus: () => void = () => {
```
You would want to add flow types for the function signature here. Not needed right now, but it'd be good to have it in mind when you add the `onFocus` and `onBlur` props later.

jandrade commented on packages/wonder-blocks-form/src/components/text-field.js:
```diff
@@ -32,23 +64,37 @@ const styles = StyleSheet.create({
+            color: Color.offBlack32,
+        },
+    },
+    focused: {
```
Nice! thanks.

jandrade commented on packages/wonder-blocks-form/src/components/text-field.js:
```diff
@@ -7,16 +7,48 @@ import Color from "@khanacademy/wonder-blocks-color";
                     styles.input,
                     typographyStyles.LabelMedium,
-                    styles.default,
+                    this.state.enabled && styles.default,
```
Based on the first comment, this would be translated to `styles.default, ....` and it's replaced with whatever UI state changes in the component (via `state` with `error, focused` or via `props` with `disabled`).

jandrade commented on packages/wonder-blocks-form/src/components/text-field.js:
```diff
@@ -7,16 +7,48 @@ import Color from "@khanacademy/wonder-blocks-color";
+        });
+    };
+
+    handleOnBlur: () => void = () => {
+        this.setState({
+            focused: false,
+        });
+    };
+
```
As you are adding `state` and handlers here, this is a good point to start adding unit tests so we can verify that some behaviors in the component work as we intend :). Happy to chat more about this part in detail.

jandrade reviewed (changes requested):
Thanks for adding more functionality on top of the first PR. Things are looking good, but I'm requesting some changes before approving:

1. Removing `enabled` from `state`.
2. Moving `disabled` to `Props`. (it'd be great if you add a new story with this prop set to `true`).
3. Adding unit tests to start verifying that the component works as expected.

RivasCVA commented:
On this new change, I moved `disabled` from a state to a prop. I removed the `enabled` state. I added a `disabled` example on Storyboard, as well as on Styleguidist. I added a few unit tests to test the states and prop.

jandrade reviewed (approved):
Looks great! Thanks for implementing the changes 💯

RivasCVA commented:
The parent pull-request ([#1123]](https://github.com/Khan/wonder-blocks/pull/1123)) has been merged into `master`, but this branch (`wb-1051`) now has conflicts with the new base branch. These conflicts must be resolved before checks can complete on this pull-request.
